### PR TITLE
Proposal for a "query_exists" function for Meson

### PR DESCRIPTION
**Pull request by rastersoft**
This function receives a string or an array of strings with paths
(relative or absolute). If at least one of the paths (which can
point to a file or a folder) does exist, it will return a string
with the first one that exists.

If none of those paths do exist, by default it will fail, showing
the list of files not available (but will return an empty string
if the "fail" keyword argument is set to False).

It is possible to exclude files (for checking only folders)
by setting the "check_files" keyword argument to False. Also
it is possible to exclude folders (for checking only files)
by setting the "check_folders" keyword argument to False.

This function can be used to check if libraries that doesn't
have a pkgconfig file are installed, by checking if their
header files exists. Since these can be stored at least at
two possible paths ("/usr/include/..." or "/usr/local/include")
it is mandatory to be able to check every possible place and
fail only if it is not available in any of them.

This function will be very useful for Autovala's Meson backend.

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -1303,6 +1304,22 @@ def func_import(self, node, args, kwargs):
     def func_files(self, node, args, kwargs):
         return [mesonlib.File.from_source_file(self.environment.source_dir, self.subdir, fname) for fname in args]
 
+    @stringArgs
+    def func_query_exists(self, node, args, kwargs):
+        check_files = kwargs.get('check_files',True)
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E231] missing whitespace after ','

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -1303,6 +1304,22 @@ def func_import(self, node, args, kwargs):
...
+        check_folders = kwargs.get('check_folders',True)
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E231] missing whitespace after ','

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -1303,6 +1304,22 @@ def func_import(self, node, args, kwargs):
...
+        for fname in args:
+            fullpath = os.path.join(self.environment.source_dir, self.subdir, fname)
+            if (check_files and os.path.isfile(fullpath)) or (check_folders and os.path.isdir(fullpath)):
+                return fname
+        if (kwargs.get('fail',True)):
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E231] missing whitespace after ','

File: mesonbuild/interpreter.py (hunk 1 above)
Comment (rastersoft) [inline mesonbuild/interpreter.py]: Ops, sorry. Sent a new patch that fixes this and the two others.

**Comment by jpakkane** [comment]
In C-like languages this check is done with `cc.has_header()`. Can this be done in Vala as well or does the compiler require you to specify full paths to said headers (or at least the directories that contain them)?

**Comment by rastersoft** [comment]
It's not a requirement for vala, but for autovala. One of the available commands is "source dependency", that specifies that the source should be built only if at least one of those files exists. I have used it in some cases to check for a header, in other for other files that I require (like some PNG icons).

**Comment by nirbheek** [comment]
I think it would be cleaner to implement this as two functions: `file_exists()` and `dir_exists()` since it makes the intent much clearer. `query_exists()` is very ambiguous and reading just the name you wouldn't guess that it checks if files/folders exist, aka "queries the filesystem". But even those names are not very nice.

I think it would be nice to have this in some form, but the name(s) need more thought. Maybe just `find_file()` and `find_directory()`.

Fwiw, till this is implemented, a workaround is to use `run_command` to run `test` in a `foreach` loop. The only problem is that it won't work on Windows. :smile:

**Comment by jpakkane** [comment]
If something like this is added it should go into a dedicated module (called, say, `io`) rather than standalone functions.

**Comment by rastersoft** [comment]
What about changing it to "check_path_exists"? (or io.check_path_exists)

**Comment by nirbheek** [comment]
If we do an `io` module, it's probably best to implement with the same names as `os.path`. So, I'd call it `io.exists`, `io.isfile`, `io.isdir`, etc.

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by jpakkane** [comment]
Closing due to inactivity. This functionality should go its own module as discussed above.

**Pull request closed without merging**